collectors/mmlsqos.go

```diff
@@ -0,0 +1,225 @@
+// Copyright 2020 Trey Dockendorf
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package collectors
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+var (
+	qosFilesystems = kingpin.Flag("collector.mmlsqos.filesystems", "Filesystems to query with mmlsqos, comma separated. Defaults to all filesystems.").Default("").String()
+	qosTimeout     = kingpin.Flag("collector.mmlsqos.timeout", "Timeout for mmlsqos execution").Default("60").Int()
+	qosSeconds     = kingpin.Flag("collector.mmlsqos.seconds", "Display the I/O performance values for the previous number of seconds. The valid range of seconds is 1-999").Default("60").Int()
+	qosMap         = map[string]string{
+		"pool":      "Pool",
+		"timeEpoch": "Time",
+		"class":     "Class",
+		"iops":      "Iops",
+		"ioql":      "Ioql",
+		"qsdl":      "Qsdl",
+		"et":        "ET",
+		"MBs":       "MBs",
+	}
+	MmlsqosExec = mmlsqos
+)
+
+type QosMetric struct {
+	Pool  string
+	Time  float64
+	Class string
+	Iops  float64
+	Ioql  float64
+	Qsdl  float64
+	ET    float64
+	MBs   float64
+}
+
+type MmlsqosCollector struct {
+	Iops   *prometheus.Desc
+	Ioql   *prometheus.Desc
+	Qsdl   *prometheus.Desc
+	ET     *prometheus.Desc
+	MBs    *prometheus.Desc
+	logger log.Logger
+}
+
+func init() {
+	registerCollector("mmlsqos", false, NewMmlsqosCollector)
+}
+
+func NewMmlsqosCollector(logger log.Logger) Collector {
+	labels := []string{"fs", "pool", "class", "measurement_period_seconds"}
+	return &MmlsqosCollector{
+		Iops: prometheus.NewDesc(prometheus.BuildFQName(namespace, "qos", "iops"),
+			"GPFS performance of the class in I/O operations per second", labels, nil),
+		Ioql: prometheus.NewDesc(prometheus.BuildFQName(namespace, "qos", "ioql"),
+			"GPFS average number of I/O requests in the class that are pending for reasons other than being queued by QoS", labels, nil),
+		Qsdl: prometheus.NewDesc(prometheus.BuildFQName(namespace, "qos", "qsdl"),
+			"GPFS average number of I/O requests in the class that are queued by QoS", labels, nil),
+		ET: prometheus.NewDesc(prometheus.BuildFQName(namespace, "qos", "et_seconds"),
```

Comment:
Is there a way to expand `et` to be a little more descriptive?

Reply:
I've changed `et` to `MeasurementInterval`